lib/app.d.ts

```diff
@@ -0,0 +1,147 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+declare namespace Realm {
+    /**
+     * The constructor of MongoDB Realm App.
+     */
+    type AppConstructor = new <FF extends FunctionsFactory>(
+        id: string,
+        configuration?: Partial<AppConfiguration>,
+    ) => App<FF>;
+
+    /**
+     * A MongoDB Realm App.
+     */
+    interface App<FF extends FunctionsFactory = FunctionsFactory> {
+        /**
+         * The id of this Realm app.
+         */
+        id: string;
+
+        /**
+         * Use this to call functions defined on the MongoDB Realm server.
+         */
+        functions: FF;
+
+        /**
+         * The last user to log in or being switched to.
+         */
+        currentUser: Realm.User | null;
+
+        /**
+         * All authenticated users.
+         */
+        allUsers: Readonly<Realm.User[]>;
+
+        /**
+         * Log in a user using a specific credential
+         * @param credentials the credentials to use when logging in
+         */
+        logIn(credentials: Credentials): Promise<Realm.User>;
+
+        /**
+         * Log out the currently authenticated user and clear any persisted authentication information.
+         */
+        logOut(): Promise<void>;
```

Comment:
I think we will only have `logOut()` on users.

Reply:
According to https://docs.google.com/document/d/1iD9ujY76BWVUOdN7rRH4XyPF9hHxx48RcGc-ZaQPyII/edit#heading=h.i99zu6swdf68 it should be available on the app instance.

Reply:
I second @kneth here. Ins't this confusing. Also (If I am not wrong) multiple users can be logged on at the same time. If true I would expect this to logout all of them!?! 
Why as a user i don't do `app.getCurrentUser.logout()`

Reply:
@blagoev we need to elevate that discussion to the entire SDK team, I'm just following the scope here. I suggest either posting a comment in the doc (tagging Jason) or posting a comment in the #realm-sdk-team channel on Slack, to get this clarified.

Reply:
Thanks for clarifying this .. I suggest I'll do a follow up PR with these changes as I fear this will ripple into too many conflicting changes for the branches I have based on this branch.